virttest/env_process.py

```diff
@@ -669,6 +677,10 @@ def preprocess(test, params, env):
     """
     error_context.context("preprocessing")
 
+    # Run this hook before any network setup stage and vm creation.
+    if preprocess_vm_off_hook:
+        preprocess_vm_off_hook(params, env)  # pylint: disable=E1102
```

Comment:
How should we handle the failures (exceptions)?

Reply:
I think handling like argument selection and implementation are all up to the caller. The VT test is already taking care of the pre- and post-processing errors and at least to my experience (already using hooks and raising LVM and other exceptions from there) these are regularly called as errors during pre- or post-processing.

Reply:
Sorry that I haven't check this comment yet since I was busy on some urgent works last few months.

Here the problem is, by the current approach, VT test does catch the pre- and post-processing errors to prevent crashing the process, but it is logically incorrect because any exception raised by hooks would break the entire pre-/post-processing, that means the codes after the hooks won't be executed anymore, which includes some important ones. It doesn't matter for the pre-processing as the test has not been started at that time, but it does affect the post- one.

Let see an example:
```python
# Run this hook after any vms are actually off to ensure data is
# written to disk.
if callable(postprocess_vm_off_hook):
    postprocess_vm_off_hook(test, params, env)  # pylint: disable=E1102

if err:
    raise RuntimeError("Failures occurred while postprocess:\n%s" % err)
```
supposes that the variable `err` is set to some valid value before `postprocess_vm_off_hook` being called, it means the test will be failed by that reason, however, what if `postprocess_vm_off_hook` raises another error? How will the test status to be set?

IMO, calling hooks are just optional steps for the two phases and others are mandatory, in any cases we should ensure the phases are finished completely.

Reply:
Thanks for elaborating, I was still thinking from the perspective of the old autotest's virttest where exceptions of any kind are always caught at singe preprocessing and postprocessing clauses. I just force pushed a version that hopefully addresses your concerns.